File: COCI '07 Contest 3 #6 Redoks.cpp
// https://dmoj.ca/problem/coci07c3p6
// Segment tree with lazy propagation
// - For each segment, we keep track of the occurrences of each digit
// - To update (and push down), we shift the digits 1 to the right, mod 10
// - To answer queries, we can get the sum of the digits using hte frequencies

#include <bits/stdc++.h>
using namespace std;

vector<int> shift(vector<int> arr, int amt) {
    amt = amt % 10;
    vector<int> result(arr.size());
    copy(arr.end() - amt, arr.end(), result.begin());
    copy(arr.begin(), arr.end() - amt, result.begin() + amt);
    return result;
}

vector<int> combine(vector<int> a1, vector<int> a2) {
    vector<int> result(a1.size());
    for (int i = 0; i < a1.size(); ++i) {
        result[i] = a1[i] + a2[i];
    }
    return result;
}

class LazySegTree {
public:
    vector<vector<int>> seg;
    vector<int> lazy;
    int size;
    vector<int> default_value;

    LazySegTree(vector<int> arr, vector<int> default_value) : default_value(default_value) {
        int layers = ceil(log2(arr.size()));
        size = 1 << layers;
        seg.resize(2 * size, default_value);
        lazy.resize(2 * size, 0);

        for (int i = 0; i < arr.size(); ++i) {
            seg[size + i][arr[i]] += 1;
        }
        for (int i = size - 1; i > 0; --i) {
            seg[i] = combine(seg[2 * i], seg[2 * i + 1]);
        }
    }

    void push_down(int i, int segment_size) {
        if (lazy[i] == 0) {
            return;
        }
        lazy[2 * i] += lazy[i];
        lazy[2 * i + 1] += lazy[i];
        seg[2 * i] = shift(seg[2 * i], lazy[i]);
        seg[2 * i + 1] = shift(seg[2 * i + 1], lazy[i]);
        lazy[i] = 0;
    }

    void update(int i, int l, int r, int cur_l, int cur_r, int diff) {
        if (cur_r < l || r < cur_l) {
            return;
        }
        int mid = (cur_l + cur_r) / 2;
        if (l <= cur_l && cur_r <= r) {
            lazy[i] += diff;
            seg[i] = shift(seg[i], diff);
            return;
        }
        push_down(i, cur_r - mid);
        update(2 * i, l, r, cur_l, mid, diff);
        update(2 * i + 1, l, r, mid + 1, cur_r, diff);
        seg[i] = combine(seg[2 * i], seg[2 * i + 1]);
    }

    vector<int> query(int i, int l, int r, int cur_l, int cur_r) {
        if (cur_r < l || r < cur_l) {
            return default_value;
        }
        if (l <= cur_l && cur_r <= r) {
            return seg[i];
        }
        int mid = (cur_l + cur_r) / 2;
        push_down(i, cur_r - mid);
        return combine(query(2 * i, l, r, cur_l, mid),
                       query(2 * i + 1, l, r, mid + 1, cur_r));
    }

    vector<int> query_range(int l, int r) {
        return query(1, l, r, 0, size - 1);
    }

    void update_range(int l, int r, int diff) {
        update(1, l, r, 0, size - 1, diff);
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, Q;
    cin >> N >> Q;
    vector<int> arr(N);
    for (int i = 0; i < N; ++i) {
        char digit;
        cin >> digit;
        arr[i] = digit - '0';
    }

    LazySegTree seg(arr, vector<int>(10, 0));

    for (int i = 0; i < Q; ++i) {
        int l, r;
        cin >> l >> r;
        l--; r--;  // convert to 0-indexing
        vector<int> freq = seg.query_range(l, r);
        int sum = 0;
        for (int i = 0; i < 10; ++i) {
            sum += freq[i] * i;
        }
        cout << sum << "\n";
        seg.update_range(l, r, 1);
    }

    return 0;
}

File: COCI '07 Contest 3 #6 Redoks.py
# MLE, PYTHON USES TOO MUCH MEMORY, CHECK C++ CODE
# https://dmoj.ca/problem/coci07c3p6
# Segment tree with lazy propagation
# - For each segment, we keep track of the occurrences of each digit
# - To update (and push down), we shift the digits 1 to the right, mod 10
# - To answer queries, we can get the sum of the digits using hte frequencies

import sys
from math import log2, ceil

shift = lambda arr, amt: arr[-(amt % 10):] + arr[:-(amt % 10)]  # [1,2,3,4], 2 --> [3,4,1,2]
combine = lambda a1, a2: [i + j for i, j in zip(a1, a2)]  # [1,1,1], [1,2,1] --> [2,3,1]  (adds digit count)


class LazySegTree:
    def __init__(self, arr, f, default) -> None:
        """create 1-indexed segment tree from 0-indexed array"""
        self.f = f  # function used to combine segments
        self.default = default
        layers = ceil(log2(N))  # max depth of the seg tree, root has depth 0
        self.size = 1 << layers
        self.seg = [default.copy() for _ in range(2 * self.size)]  # 1-indexed, need an extra layer for the actual data
        self.lazy = [0] * (2 * self.size)

        for i in range(N):  # base layer
            self.seg[self.size + i][arr[i]] += 1
        for i in reversed(range(1, self.size)):  # create other layers from base
            self.seg[i] = f(self.seg[i * 2], self.seg[i * 2 + 1])

    def push_down(self, i, segment_size):
        """transfer lazy tag from parent to children"""
        if self.lazy[i] == 0:
            return
        self.lazy[i * 2] += self.lazy[i]  # transfer lazy tags
        self.lazy[i * 2 + 1] += self.lazy[i]
        self.seg[i * 2] = shift(self.seg[i * 2], self.lazy[i])  # update value of segments
        self.seg[i * 2 + 1] = shift(self.seg[i * 2 + 1], self.lazy[i])
        self.lazy[i] = 0

    def update(self, i, l, r, cur_l, cur_r, diff) -> None:
        """add diff to [left, right]"""
        if cur_r < l or r < cur_l:  # fully outside segment
            return
        mid = (cur_l + cur_r) // 2
        if l <= cur_l and cur_r <= r:  # fully inside segment: update lazily
            self.lazy[i] += diff
            self.seg[i] = shift(self.seg[i], diff)
            return
        self.push_down(i, cur_r - mid)  # partial covers, recurse deeper
        self.update(i * 2, l, r, cur_l, mid, diff)
        self.update(i * 2 + 1, l, r, mid + 1, cur_r, diff)
        self.seg[i] = self.f(self.seg[i * 2], self.seg[i * 2 + 1])

    def query(self, i, l, r, cur_l, cur_r) -> int:
        """query [left, right]"""
        if cur_r < l or r < cur_l:  # fully outside segment
            return self.default
        if l <= cur_l and cur_r <= r:  # fully inside segment
            return self.seg[i]

        mid = (cur_l + cur_r) // 2  # traverse down further
        self.push_down(i, cur_r - mid)
        return self.f(self.query(i * 2, l, r, cur_l, mid),
                      self.query(i * 2 + 1, l, r, mid + 1, cur_r))

    def query_range(self, l, r):
        return self.query(1, l, r, 0, self.size - 1)

    def update_range(self, l, r, diff):
        return self.update(1, l, r, 0, self.size - 1, diff)


input = sys.stdin.readline
N, Q = map(int, input().split())
arr = list(map(int, list(input().strip())))

seg = LazySegTree(arr, f=combine, default=[0] * 10)

for _ in range(Q):
    l, r = map(lambda x: int(x) - 1, input().split())  # convert to 0-indexing
    freq = seg.query_range(l, r)
    print(sum(freq[i] * i for i in range(10)))  # sum of numbers on dials

    seg.update_range(l, r, 1)
